### ISSUE TITLE
Feature Request: Ligature Support

### ISSUE THREAD

**Issue by bojjenclon** [pkulchenko/ZeroBraneStudio#960]
I use the Fira Code font for coding, and part of the reason I love it is its use of ligatures. Most editors (such as VS Code) have a toggleable option to enable/disable ligatures. I don't see any such option in the user config sample, and ligatures don't seem to work out of the box when setting Fira Code as the default font.

**Comment by pkulchenko** [comment]
This is not supported in the current version of the IDE, as it was not supported in wxwidgets this version is built with. The support for ligatures on Windows has already been added (with the work for the ticket I opened two years ago https://trac.wxwidgets.org/ticket/17804), but you'll have to compile the binaries to get access to the new functionality or wait for the next wxwidgets upgrade in the IDE.

**Comment by pkulchenko** [comment]
> How is the wxwidgets dependency managed in this project? Is it as simple as building the latest wxwidget source into a dll and replacing the existing dll in bin/clibs, or is it more nuanced than that?

It's a bit more nuanced than that, as using wxwidgets requires wxlua wrapper, so there are two dependencies. ZBS source includes builds scripts that will retrieve, build, and install wxwidgets and wxlua libraries, but the current wxlua version doesn't work with the most recent wxwidgets code due to some API changes in wxwidgets. I've been waiting for [wxiwdgets 3.1.2 release](https://www.wxwidgets.org/news/2018/12/wxwidgets-3.1.2-released/) to bring wxlua in sync with the most recent wxwidgets version.

> Should I assume there are potentially breaking API changes that have blocked the upgrade of the DLL on your side up to this point?

There are some wxwidgets API changes that need to be taken care of in wxlua, but there is nothing on the IDE side that needs to be adjusted to accept wxlua library to make it work. In fact, the IDE still maintains compatibility with wxlua/wxwidgets 2.9.5+ versions (although some functionality may not be available).

If you really want to give it a try yourself and don't want to wait for my changes, you can just pick some wxwidgets from earlier this year (before the API changes were introduced) and compile it using the build scripts included with the IDE. I can provide instructions on how this can be done.

**Comment by bojjenclon** [comment]
Thank you for the detailed response. I don't necessarily have a problem with waiting, I just wasn't sure where the upgrade was on your radar (in other words, if you expected it to happen within a few months or if its a long term change). Its nice to see you're so active in the development of this, a lot of projects like this tend to fall to the wayside after a while, but I've watched ZBS grow for quite a while.

**Comment by pkulchenko** [comment]
Yes, the upgrade is definitely planned (although I'm likely to wait for some of scintilla changes to be integrated into wxwidgets), but at the very least I'll update wxlua so that it compiles and works with wxwidgetds 3.1.2 to allow the provided build scripts to build the most recent wxwidgets version.

Thank you for the feedback!

**Comment by pkulchenko** [comment]
@bojjenclon, can you check the binaries from #999 for ligatures support?

**Comment by bojjenclon** [comment]
So sorry about the wait!

I saw commit cabea9f was on branch wxwidgets-update-313, so I DL'd the repo at that branch and tried running with a ligature enabled font (Fira Code). I don't see them appearing with things like == or => as I do in other editors. Is there a step or a flag I'm missing perhaps?

**Comment by pkulchenko** [comment]
@bojjenclon, you are correct; it turned out that d2d support that is needed for ligatures has to be explicitly enabled during the build process (and the compiler needs to have d2d headers). I've successfully compiled and tested the changes, so should have an update pushed soon.

**Comment by pkulchenko** [comment]
@bojjenclon, I pushed the changes to https://github.com/pkulchenko/ZeroBraneStudio/tree/direct2d branch. You'll need to configure the font (for example, `editor.fontname="Fira Code"`) and then execute `ide:GetEditor():SetTechnology(1)` from the console, which will then enable ligatures support in the current editor tab. I'm still working on a more complete changes (there is going to be a configuration setting), but wanted to give you a chance to check how it works and to see if you run into any issues with it.

**Comment by moteus** [comment]
It is possible to turn on this setting via simple package [like this](https://github.com/moteus/zbs-package/blob/master/zbstudio/packages/endatlast.lua).

**Comment by bojjenclon** [comment]
> 
> 
> @bojjenclon, I pushed the changes to https://github.com/pkulchenko/ZeroBraneStudio/tree/direct2d branch. You'll need to configure the font (for example, `editor.fontname="Fira Code"`) and then execute `ide:GetEditor():SetTechnology(1)` from the console, which will then enable ligatures support in the current editor tab. I'm still working on a more complete changes (there is going to be a configuration setting), but wanted to give you a chance to check how it works and to see if you run into any issues with it.

I can confirm it works!

**Comment by bojjenclon** [comment]
One thing I'm noticing already is a lot of tearing. Scrolling seems to sometimes cause tearing which will randomly disappear during said scroll.

**Comment by pkulchenko** [comment]
@bojjenclon, do you happen to have a screenshot? I tried scrolling, but haven't seen any problems with it. The original ticket (https://trac.wxwidgets.org/ticket/17804) included quite a bit of discussion on visual and performance aspects related to the changes, so there may still be something going on. Also, what Windows version are you on?

**Comment by pkulchenko** [comment]
> It is possible to turn on this setting via simple package like this.

@moteus, agree, but I plan to have a config setting for it, so it should be easy to enable.

**Comment by pkulchenko** [comment]
@bojjenclon, I have to put this change on hold, as I noticed an issue with the auto-complete popup (it's larger with the SetTechnology call). I updated the original wxwidgets ticket (https://trac.wxwidgets.org/ticket/17804#comment:32) and will wait for some update. It's likely that this change won't go into this update unfortunately (you can still use the binary provided, but I probably won't be merging it into master until the issue is addressed).

**Comment by pkulchenko** [comment]
Just to note that a patch for wxwidgets is required to enable d2d support during the compilation (in addition to using a version of gcc with d2d headers):

```diff
diff --git a/setup.h.in b/setup.h.in
index b33699b9c9..3ff80a19c1 100644
--- a/setup.h.in
+++ b/setup.h.in
@@ -674,7 +674,7 @@
 #if defined(_MSC_VER) && _MSC_VER >= 1600
     #define wxUSE_GRAPHICS_DIRECT2D wxUSE_GRAPHICS_CONTEXT
 #else
-    #define wxUSE_GRAPHICS_DIRECT2D 0
+    #define wxUSE_GRAPHICS_DIRECT2D wxUSE_GRAPHICS_CONTEXT
 #endif
```

**Comment by pkulchenko** [comment]
@bojjenclon, @moteus, this has been implemented and included in all-binaries-190c branch. You'd need to set `editor.showligatures=true` to enable.

**Comment by bojjenclon** [comment]
Sorry about taking so long to get back to this! I just downloaded branch upgrade-binaries-190d, set editor.showligatures = true in my config, and lo and behold I see working ligatures! I think you can mark this as working. :)

I will say though I checked your latest commit that references this issue and I think you have a flag mismatch.
In src/config.lua you set ligatures = false, but then in src/editor/editor.lua you check for the state of edcfg.showligatures.

**Comment by pkulchenko** [comment]
> In src/config.lua you set ligatures = false, but then in src/editor/editor.lua you check for the state of edcfg.showligatures.

Good catch; I'll fix this. Thank you for testing the changes!

**Comment by pkulchenko** [comment]
This has been implemented as part of the changes in #999 that have been merged.
